internal/porchutil/function.go

```diff
@@ -0,0 +1,155 @@
+package porchutil
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/kubernetes"
+
+	// Enable the GCP Authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const expirationNanoSecond = 3
+
+var (
+	functionAPI     = "/apis/" + v1alpha1.SchemeGroupVersion.String() + "/functions"
+	PorchExpiration = time.Duration(expirationNanoSecond) * time.Second
+)
+
+// NewPorchFunctionClient provides the PorchFunctionClient which connects to Porch server
+// and gets the functions.
+func NewPorchFunctionClient() *PorchFunctionClient {
+	functions, err := visitFunctions()
+	if err != nil {
+		return &PorchFunctionClient{IsPorchAvailable: false, Functions: nil}
+	}
+	return &PorchFunctionClient{IsPorchAvailable: true, Functions: functions}
+}
+
+type PorchFunctionClient struct {
+	IsPorchAvailable bool
+	Functions        []v1alpha1.Function
+}
+
+// IsFunctionTypeMatch determines whether the `function` (which can be multi-typed), belongs
+// to the expectedType. type value should only be `validator` or `mutator`.
+func IsFunctionTypeMatch(function v1alpha1.Function, expectedType string) bool {
+	for _, actualType := range function.Spec.FunctionTypes {
+		if string(actualType) == expectedType {
+			return true
+		}
+	}
+	return false
+}
+
+// ContainsKeywords determines whether the `function` has keywords which match the `expectedKeywords`.
+// Experimental: This logic may change to only if all function keywords are found from  expectedKeywords,
+// can it claims a match (return true).
+func ContainsKeywords(function v1alpha1.Function, expectedKeywords []string) bool {
+	if len(expectedKeywords) == 0 {
+		// Accept all functions if keywords are not given.
+		return true
+	}
+	for _, actual := range function.Spec.Keywords {
+		for _, expected := range expectedKeywords {
+			if actual != expected {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// ListKeywords returns all the unique keywords from Porch Functions.
+func (p *PorchFunctionClient) ListKeywords(cmd *cobra.Command) ([]string, error) {
+	keywordsMap := map[string]bool{}
+	fnType := cmd.Flag("type").Value.String()
+	for _, function := range p.Functions {
+		if !IsFunctionTypeMatch(function, fnType) {
+			continue
+		}
+		for _, keyword := range function.Spec.Keywords {
+			keywordsMap[keyword] = true
+		}
+	}
+	var keywords []string
+	for keyword := range keywordsMap {
+		keywords = append(keywords, keyword)
+	}
+	return keywords, nil
+}
+
+func getKeywordsFromFlag(cmd *cobra.Command) []string {
+	flagVal := cmd.Flag("keywords").Value.String()
```

Comment:
Cobra doesn't have facility to accept lists as flag values? I'm just wondering if we can avoid the custom parser.

Reply:
Yeah, I explored this a little bit. I used `StringArrayVarP` for "keywords", which users can provide an array of keywords in the form as `--keywords XX --keywords YY`. But when parsing the flag from commands, it does not provide such options.

Reply:
I see. re-reading the code it seems that for the completion, less arg parsing has been done. That _may_ mean that if customer enters --keywords XX --keywords YY ... the code may not see both options?